src/compiler/checker.ts

```diff
@@ -4723,21 +4726,32 @@ module ts {
 
         function getInferredType(context: InferenceContext, index: number): Type {
             let inferredType = context.inferredTypes[index];
+            let inferenceSucceeded: boolean;
             if (!inferredType) {
                 let inferences = getInferenceCandidates(context, index);
                 if (inferences.length) {
                     // Infer widened union or supertype, or the undefined type for no common supertype
                     let unionOrSuperType = context.inferUnionTypes ? getUnionType(inferences) : getCommonSupertype(inferences);
-                    inferredType = unionOrSuperType ? getWidenedType(unionOrSuperType) : inferenceFailureType;
+                    inferredType = unionOrSuperType ? getWidenedType(unionOrSuperType) : unknownType;
+                    inferenceSucceeded = !!unionOrSuperType;
                 }
                 else {
                     // Infer the empty object type when no inferences were made
                     inferredType = emptyObjectType;
+                    inferenceSucceeded = true;

```

Comment:
It's not yet a PR; I'd like to make it a PR if we could solve some of the original concerns in #360.

To clarify, when inference for a type argument fails, the process falls back to using `{}` as a type argument. :+1: for explaining what a failure scenario would be in contrast to no inferences being found.